Lesson5 try1

IgushkinV opened the pull request:
ДЗ №5 IO. Первая попытка сдачи.

IgushkinV committed:
Промежуточный коммит.
Реализована возможность добавления строк в файл, чтение сток из файла.

IgushkinV committed:
Промежуточный коммит.
Реализовано удаление из файла. Осталось подчистить.

IgushkinV committed:
Промежуточный коммит. Добавлен недостающий файл.

IgushkinV committed:
Итоговый коммит.
Вынесены методы для работы с файлами в отдельный класс. Методы сделал static.
Написал javaDoc для всех методов и классов.

tlt-java-courses commented on src/main/java/Igushkin/Homeworks/Lesson5/CommandHandler.java:
```diff
@@ -0,0 +1,5 @@
+package Igushkin.Homeworks.Lesson5;
+
+public interface CommandHandler {
+    public void handle() ;
```
А может мы попробуем принимать команду на вход?

IgushkinV replied:
Сделано.

tlt-java-courses commented on src/main/java/Igushkin/Homeworks/Lesson5/FileMethods.java:
```diff
@@ -0,0 +1,98 @@
+        //2 варианта записи
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(fileName))) {
+            //1 вариант - запись между существующих строк. Отсчет строк с 0.
+            if (lineNumber < linesFromFile.size()) {
```
Может не будем делать один божественный метод, а разнесем на 2 отдельных чтобы не усложнять?

IgushkinV replied:
Разнес на 2 метода. Один добавляет в середину файла, другой после существующих строк, вставляя пустые. 
Такие, специфичные методы получились.

IgushkinV replied:
У меня один и тот же набор параметров, но в зависимости от значения одного из них я вызываю один из двух методов. Это ок?

tlt-java-courses commented on src/main/java/Igushkin/Homeworks/Lesson5/Filter.java:
```diff
@@ -0,0 +1,119 @@
+    /**
+     * Поле. Содержит массив параметров, полученных из переданной строки.
+     */
+    private String[] params;
```
А давай класс не будет хранить у себя состояние?

IgushkinV replied:
Переписал.

tlt-java-courses commented on src/main/java/Igushkin/Homeworks/Lesson5/Filter.java:
```diff
@@ -0,0 +1,119 @@
+        this.line = line;
+        this.params = line.split(" ");
+        this.command = params[0];
+        switch (command) {
```
А мы можем как-нибудь обойтись без switch? Вспомни про юзеров и их роли.

IgushkinV replied:
Переписал, используя HashMap.

tlt-java-courses commented on src/main/java/Igushkin/Homeworks/Lesson5/Filter.java:
```diff
@@ -0,0 +1,119 @@
+        this.command = params[0];
+        switch (command) {
+            case "add":
+                String[] halfsOfInputLine = line.split("\"",2);
```
Давай обработчики сами будут решать смогут ли они распарсить команду или нет?

IgushkinV replied:
Переписал. Разнес код из класса Filter по соответствующим обработчикам.

tlt-java-courses commented on src/main/java/Igushkin/Homeworks/Lesson5/Filter.java:
```diff
@@ -0,0 +1,119 @@
+                            log.warn("Введен отрицательный номер строки, используйте неотрицательный");
+                            break;
+                        }
+                        commandHandler = new MyWriter(lineNumber, params[2], stringToAdd);
```
Может создадим обработчиков в классе 1 раз и будем их переиспользовать?

IgushkinV replied:
Сделал. Обработчики создаются при создании Filter.

tlt-java-courses commented on src/main/java/Igushkin/Homeworks/Lesson5/MyDeleter.java:
```diff
@@ -0,0 +1,56 @@
+@Slf4j
+public class MyDeleter implements CommandHandler {
+
+    int lineNumber;
```
А может обработчики не будут хранить у себя состояние?

IgushkinV replied:
Переписал.

tlt-java-courses commented on src/main/java/Igushkin/Homeworks/Lesson5/MyDeleter.java:
```diff
@@ -0,0 +1,56 @@
+     * Удалаяет конкретную строку либо последнюю строку файла в соответствии с переданными в конструктор параметрами.
+     */
+    @Override
+    public void handle() {
```
Давай сделаем так чтобы они принимали текстовую команду?

IgushkinV replied:
Переписал.

IgushkinV committed:
Исправление замечаний к Первой попытке сдачи.
Теперь классы Filter и классы обработчиков команд не хранят свое состояние.
Убрал использование switch для выбора обработчика.
Теперь обработчики принимают текст запроса пользователя и сами проверяют корректность переданной команды.
Теперь обработчики создаются один раз, потом переиспользуются.
Вместо метода stringWriter создал два метода для записи внутрь файла и после существующих строк файла.

tlt-java-courses commented on src/main/java/Igushkin/Homeworks/Lesson5/Filter.java:
```diff
@@ -0,0 +1,42 @@
+    private HashMap<String, CommandHandler> handlersMap = new HashMap<>();
+
+    public Filter() {
+        this.handlersMap.put("add", new MyWriter());
```
Сейчас не сильно критично, но мы можем попробовать ограничить набор команд при помощи enum

IgushkinV replied:
Попробовал ограничить. Пришлось усложнить код. Пожалуй, в проекте, где будут другие строковые значения на вход, это будет оправдано.

tlt-java-courses commented on src/main/java/Igushkin/Homeworks/Lesson5/Filter.java:
```diff
@@ -0,0 +1,42 @@
+    public void process(String line) {
+        String[] params = line.split(" ");
+        String command = params[0];
+        if (handlersMap.containsKey(command)) {
```
у мапы есть такой метод
https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#getOrDefault-java.lang.Object-V-
 тут можно было бы подумать в сторону обработчика несуществующих команд.

IgushkinV replied:
Добавил обработчик для несуществующих команд.

IgushkinV committed:
Исправление замечаний №2 к ДЗ.
Добавил enum Command, написал в нем метод getCommandByString (String str).
Переделал обработку несуществующей команды с помощью Map.getOrDefault. Добавил для этого еще один обработчик.

IgushkinV committed:
Добавление к последнему коммиту. Переписал, чтобы было "один раз создали обработчик, потом его используем".
Добавил в фильтре проверку на строку вида " " (вызывала эксепшн).
